dbms/src/Storages/MergeTree/MergeTreePartition.cpp

```diff
@@ -0,0 +1,128 @@
+#include <Storages/MergeTree/MergeTreePartition.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Storages/MergeTree/MergeTreeDataPart.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/HashingWriteBuffer.h>
+#include <Core/FieldVisitors.h>
+#include <DataTypes/DataTypeDate.h>
+#include <Common/SipHash.h>
+#include <Common/typeid_cast.h>
+#include <Common/hex.h>
+
+#include <Poco/File.h>
+
+namespace DB
+{
+
+static ReadBufferFromFile openForReading(const String & path)
+{
+    return ReadBufferFromFile(path, std::min(static_cast<Poco::File::FileSize>(DBMS_DEFAULT_BUFFER_SIZE), Poco::File(path).getSize()));
+}
+
+String MergeTreePartition::getID(const MergeTreeData & storage) const
+{
+    if (value.size() != storage.partition_expr_columns.size())
+        throw Exception("Invalid partition key size: " + toString(value.size()), ErrorCodes::LOGICAL_ERROR);
+
+    if (value.empty())
+        return "all";
+
+    /// In case all partition fields are represented by integral types, try to produce a human-readable partition id.
+    /// Otherwise use a hex-encoded hash.
+
+    bool are_all_integral = true;
+    for (const Field & field : value)
+    {
+        if (field.getType() != Field::Types::UInt64 && field.getType() != Field::Types::Int64)
+        {
+            are_all_integral = false;
+            break;
+        }
+    }
+
+    String result;
+
+    if (are_all_integral)
+    {
+        FieldVisitorToString to_string_visitor;
+        for (size_t i = 0; i < value.size(); ++i)
+        {
+            if (i > 0)
+                result += '-';
+
+            if (typeid_cast<const DataTypeDate *>(storage.partition_expr_column_types[i].get()))
+                result += toString(DateLUT::instance().toNumYYYYMMDD(DayNum_t(value[i].get<UInt64>())));
+            else
+                result += applyVisitor(to_string_visitor, value[i]);
+        }
```

Comment:
I assume that you had the same thoughts while writing this code:
"we cannot do the same for DateTime because it will depend on timezones".

Better to write it in comment.

Reply:
Actually no, I was just lazy :) Good point.